### PR TITLE
Raft pre elections should be compatible with refuse_to_be_leader

### PR DESCRIPTION
**Pull request by andrewkerr9000**

### CODE REVIEW

**Comment by andrewkerr9000** [comment]
I'll rearrange code a bit to make the diff clearer and get rid of the TODOs. Also fix committer name.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java [hunk 1]
```diff
@@ -167,49 +192,127 @@ public Outcome handle( RaftMessages.NewEntry.BatchRequest batchRequest ) throws
...
+        private static ElectionTimeoutHandler instance = new PreVoteUnsupportedHandler();
     }
 
-    class PreVoteActiveHandler extends PreVoteSupportedHandler
+    private static class PreVoteUnsupportedRefusesToLead implements ElectionTimeoutHandler
     {
+        @Override
+        public Outcome handle( RaftMessages.Timeout.Election election, Outcome outcome, ReadableRaftState ctx, Log log ) throws IOException
+        {
+            log.info( "Election timeout triggered but refusing to be leader" );
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java:247]: This will be very useful in debugging!

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java [hunk 2]
```diff
@@ -167,49 +192,127 @@ public Outcome handle( RaftMessages.NewEntry.BatchRequest batchRequest ) throws
...
+    {
+        @Override
+        public Outcome handle( RaftMessages.Timeout.Election election, Outcome outcome, ReadableRaftState ctx, Log log ) throws IOException
         {
-            super( ctx, log );
+            log.info( "Election timeout triggered but refusing to be leader" );
+            Set<MemberId> memberIds = ctx.votingMembers();
+            if ( memberIds != null && memberIds.contains( ctx.myself() ) )
+            {
+                outcome.setPreElection( true );
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java:263]: Why are we manipulating this if we are anyway refusing to be leader?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java [hunk 3]
```diff
@@ -81,17 +81,24 @@ public Outcome handle( RaftMessages.RaftMessage message, ReadableRaftState ctx,
...
+        private final ElectionTimeoutHandler electionTimeoutHandler;
 
-        Handler( ReadableRaftState ctx, Log log )
+        Handler( PreVoteRequestHandler preVoteRequestHandlerDelegate, PreVoteResponseHandler preVoteResponseHandler,
+                ElectionTimeoutHandler electionTimeoutHandler, ReadableRaftState ctx, Log log )
         {
             this.ctx = ctx;
             this.log = log;
             this.outcome = new Outcome( FOLLOWER, ctx );
+            this.preVoteRequestHandlerDelegate = preVoteRequestHandlerDelegate;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java]: Is there a reason to single this one out with Delegate in the name? Hard for me to see without IDE-support, but can't really notice anything special through cursory browsing here.

**Comment by martinfurmanski** [comment]
I had just very minor comments. I think the functionality looks good enough for a merge, however the PR run seems to have failed miserably. Don't know if just flaky or perhaps some lifecycle issue in test refactoring.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java (hunk 3 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java]: Probably an oversight from an earlier version.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java (hunk 2 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java:263]: `preElection` needs to be true to participate in pre elections (`PreVoteRequestVotingHandler`). However it won't send pre vote requests and it will ignore pre vote responses.

See lines 361-376.

**Comment by andrewkerr9000** [comment]
Looks like a PreElectionIT is failing, and causing a cascade of failures by not cleaning up after itself.

**Comment by andrewkerr9000** [comment]
@martinfurmanski Tests should be stable now.

tl;dr: we need to prevent members that refuse to be leader bootstrapping the cluster. HazelcastDiscoveryService already did this, but SharedDiscoveryService did not. Unfortunately the logic is coupled to implementation specific data structures and cannot easily be extracted further up the call hierarchy.

**Comment by andrewkerr9000** [comment]
Cypher test failed, rebased.

**Review by martinfurmanski — APPROVED** [review]
